pystac_client/warnings.py

```diff
@@ -0,0 +1,73 @@
+from contextlib import contextmanager
+import warnings
+
+
+FALLBACK_MSG = "Falling back to pystac. This might be slow."
+
+
+class PystacClientWarning(UserWarning):
+    """Base warning class"""
+
+    ...
+
+
+class NoConformsTo(PystacClientWarning):
+    """Inform user when client does not have "conformsTo" set"""
+
+    ...
+
+
+class DoesNotConformTo(PystacClientWarning):
+    """Inform user when client does not conform to extension"""
+
+    ...
+
+
+class MissingLink(PystacClientWarning):
+    """Inform user when link is properly implemented"""
+
+    ...
+
+
+class FallbackToPystac(PystacClientWarning):
+    """Inform user when falling back to pystac implementation"""
+
+    ...
+
+
+@contextmanager
+def strict():  # type: ignore
+    """Context manager for raising all pystac-client warnings as errors
+
+    For more fine-grained control or to set filter warnings in the whole
+    python session, use the ``warnings`` module directly.
+
+    Examples:
+
+    >>> import warnings
+    ... warnings.filterwarnings("error", category=FallbackToPystac)
+    """
+    warnings.filterwarnings("error", category=PystacClientWarning)
+    try:
+        yield
+    finally:
+        warnings.resetwarnings()
```

Comment:
I'm a little wary about this. `resetwarnings` returns warnings to their default state, which they might not be in when we call `strict` (or whatever). So we could be changing the warning setup without the user knowing.

It'd be good to use https://docs.python.org/3/library/warnings.html#warnings.catch_warnings, but it's single-threaded only ... we'd have to call that out explicitly in the docs.

Reply:
Yeah I am kind of so-so on these helper functions in general. I doubt what I have here is thread-safe.

Reply:
I changed this to set the filter back to default just for our special class. Does that feel better?

Reply:
Yes I like that a lot, thanks!